text/3487-feature-visibility.md

```diff
@@ -0,0 +1,160 @@
+feature as non-public API. This is described as the following:
+
+- The feature is always usable within the same crate:
+  - Enabled by other features, e.g.
+    `foo = { enables = [some-private-feature] }`, is allowed
+  - Referenced in `[[bench]]` and `[[test]]` target `required-features`
+  - Using the feature on the command-line is allowed
+- Users may explicitly specifying the private features for their dependencies
+  on the command-line (e.g. `--features somecrate/private-feature`) which would
+  otherwise be forbidden
+- The feature should not be accepted by `cargo add --features`
+- The feature should not be reported from `cargo add`'s feature output report
+  - A future tool like `cargo info` shouldn't display information about these
+    features
+- Once `rustdoc` is able to consume feature metadata, `rustdoc` should not
+  document these features unless `--document-private-items` is specified
+
+Attempting to use a private feature in any of the forbidden cases should result
+in an error. Exact details of how features work will likely be refined during
+implementation and experimentation.
+
+Two sample use cases for `public = false` include:
+
+- `docs.rs` having a way to know which features should be hidden
+- Features that are included in feature chains (feature `a` enables feature `b`)
+  but not meant for public consumption could be marked not public
+
+This feature requires adjustments to the index for full support. This RFC
+proposes that it would be acceptable for the first implementation to simply
+strip private features from the manifest; this meanss that there will be no way
+to `cfg` based on these features.
+
+Full support does not need to happen immediately, since it will require this
+information be present in the index. [Index changes] describes how this can take
+place.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Added complexity to Cargo. Parsing is trivial, but exact implementation
+  details do add test surface area
+- Added Cargo arguments if escape hatches for `public` are created
+- `public` uses may not be common enough to be worth including
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+TODO
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- `docs.rs` treats features with a leading `_` as private / hidden
+- Ivy has a [visibility attribute] for its configuration (mentioned in
+  [cargo #10882])
+- Discussion on stable/unstable/nightly-only features
+  <https://github.com/rust-lang/cargo/issues/10881>
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- Are the semantics of `public` proposed in this RFC suitable? Should private
+  features be usable in examples or integration tests without a `--features`
+  argument?
+- Should there be a simple `--allow-private-features` flag that allows using all
+  features, such as for crater runs? This can be decided during implementation.
```

Comment:
Why would crater runs need this?

I feel like it'd be a bad precedence for us to create a flag for overriding visibility.  We also don't have one for source code.

Reply:
I removed this, I think at some point I was thinking that tests wouldn't have private features usable by default so crater runs would fail.